### PR TITLE
feat(hint): added hasNoOffset prop

### LINKED ISSUE #10485: Hint - add hasNoOffset support to actions
In https://github.com/patternfly/patternfly/pull/6708, we added `.pf-m-no-offset` to `.pf-v6-c-hint__actions`, so we'll need support for that in the react `<Hint>`. This basically matches the same feature for `<Card>` actions, so maybe the API can be the same? https://github.com/patternfly/patternfly-react/blob/e4debe98bdcfacfaefe8e0f2c4f7799421bb35d0/packages/react-core/src/components/Card/CardActions.tsx#L10-L11

### PR DESCRIPTION
**Pull request by evwilkin**
**What**: Closes #10485 

**Additional issues**:  Follow-up to https://github.com/patternfly/patternfly/pull/6708

This PR:
- bumps the core version
- adds support & tests for `hasNoOffset` prop on the Hint component
- updates the existing `actions` behavior to only conditionally render the `div.pf-v6-c-hint__actions` if the `actions` prop is passed

### CODE REVIEW

File: packages/react-core/src/components/Hint/Hint.tsx [hunk 1]
```diff
@@ -9,11 +9,19 @@ export interface HintProps {
   className?: string;
   /** Actions of the hint. */
   actions?: React.ReactNode;
+  /** Flag indicating that the actions have no offset */
+  hasNoOffset?: boolean;
```
Comment (tlabaj) [inline packages/react-core/src/components/Hint/Hint.tsx:13]: Is there a case where we would **not** want the Hint with actions to have the `pf-m-no-offset` applied?  It is should always be applied with action to get the layout correct, maybe we should just apply it automatically when we have actions?

What do you think @mcoker ?

Comment (tlabaj) [reply packages/react-core/src/components/Hint/Hint.tsx:13]: So I looked back at he card's original  implementation of this feature, if the use case is similar, just ignore my comment above.

Comment (mcoker) [reply packages/react-core/src/components/Hint/Hint.tsx:13]: tl;dr - IMO the default should just match the card. `hasNoOffset` (aka `.pf-m-no-offset`) should be an opt-in to to handle edge cases.

But here's what I know - the primary use case for hint actions (and same with `<Card>` and its actions) is a plain action(s) (like a kebab dropdown/toggle), and since there is no border/boundary on the plain action to show its bounding box, we offset the hint/card actions element up vertically by the padding value of a plain action, which visually aligns the plain action's icon/text with the hint/card title. It even works well for non-plain things. Another cool thing the default layout does is keeps actions in the card/hint from making the card/hint's header element taller based off of the height of the action. That means 2 cards/hints can be placed side-by-side, one has an action in the header and the other doesn't, and the header height is the same between the two - meaning the header text and body content are aligned and have a similar layout.

`hasNoOffset` is more of an edge case variation for when our default vertical offset messes up the layout. There is an example in core that has a good use case for `hasNoOffset`, when 1) the card title is larger than the default, and 2) you're using a non-plain action - https://www.patternfly.org/components/card/html#actions-with-no-offset

Here are some screenshots:

### Default layout - looks  👍
Screenshot 2024-06-03 at 3 16 42 PM

### Border around the plain action and the available space in the header to show what the default offset does
Screenshot 2024-06-03 at 3 18 04 PM

### [Core example using `hasNoOffset`](https://www.patternfly.org/components/card/html#actions-with-no-offset)
Screenshot 2024-06-03 at 3 19 26 PM

### What the above example looks like if you remove `hasNoOffset`
Screenshot 2024-06-03 at 3 19 43 PM
Screenshot 2024-06-03 at 3 19 46 PM

And here is what it looks like when 2 cards/hints are side-by-side and one has an action and the other doesn't. Because of the default offset, the presence of the action doesn't change the header height, so they line up nicely beside one another

Screenshot 2024-06-03 at 3 37 14 PM

And here is what it looks like with `hasNoOffset` on the card with the action. You can see the titles and body text aren't aligned because the action changes the header height

Screenshot 2024-06-03 at 3 37 29 PM

**Review by mcoker — APPROVED** [review]
LGTM! One comment would just be that where the prop goes is different between the card and hint and if it goes on the top-level component, I wonder if the prop name should reference "actions" (eg, `hasNoActionsOffset`).

File: packages/react-core/src/components/Hint/examples/Hint.md [hunk 2]
```diff
@@ -20,3 +20,15 @@ import EllipsisVIcon from '@patternfly/react-icons/dist/esm/icons/ellipsis-v-ico
...
 
 ```
+
+### Basic without no actions offset
+
+The `actions` property for `<Hint>` can optionally be combined with the `hasNoOffset` property, which is `false` by default. When `hasNoOffset` is `false`, a negative margin is applied to help align default-sized hint titles with hint actions.
+
+You may use `hasNoOffset` to remove this negative margin, which better aligns hint actions in implementations that use large hint titles, for example.
+
+Select the "actions hasNoOffset" checkbox in the example below to illustrate this behavior.
```
Comment (edonehoo) [inline packages/react-core/src/components/Hint/examples/Hint.md:30]:
```suggestion
You can add actions to a hint to allow users to respond to suggestions and initiate any related processes.

When a `<Hint>` contains `actions`, use the `hasNoOffset` property to ensure that there is proper alignment between the hint title and the actions component (such as a button). For default-sized titles, pass in the `hasNoOffset` property (set to `false` by default) to apply a negative margin to the hint. For larger-sized titles, remove this negative margin by setting `hasNoOffset` to `true`.

To illustrate this behavior, select the "actions hasNoOffset" checkbox in the following example.
```
wdyt? too wordy?

File: packages/react-core/src/components/Hint/examples/Hint.md [hunk 3]
```diff
@@ -20,3 +20,15 @@ import EllipsisVIcon from '@patternfly/react-icons/dist/esm/icons/ellipsis-v-ico
 ```ts file="HintBasicWithoutTitle.tsx"
 
 ```
+
+### Basic without no actions offset
```
Comment (edonehoo) [inline packages/react-core/src/components/Hint/examples/Hint.md:24]:
```suggestion
### With actions
```
Do you think this could just be called "with actions", since that's the ultimate concept that the example is supporting? Or is it important to specifically call out the prop here

**Review by edonehoo — CHANGES_REQUESTED** [review]
some ideas! lmk what you think

File: packages/react-core/src/components/Hint/examples/Hint.md (hunk 3 above)
Comment (evwilkin) [reply packages/react-core/src/components/Hint/examples/Hint.md:24]: I think this makes sense 👍

File: packages/react-core/src/components/Hint/examples/Hint.md (hunk 2 above)
Comment (evwilkin) [reply packages/react-core/src/components/Hint/examples/Hint.md:30]: This is a confusing prop so I think the description helps, even if lengthy.  The only change I'd propose is this part:

> For default-sized titles, pass in the `hasNoOffset` property (set to `false` by default) to apply a negative margin to the hint.

The `hasNoOffset` (potentially renaming to `hasNoActionsOffset`) prop is false by default like you mention, which means that for default-sized titles the user doesn't need to think about it at all or do anything.  For this reason, I think we can call that out or simply remove this sentence and only point out how to use the prop if they actually want to change the default behavior.

Comment (evwilkin) [reply packages/react-core/src/components/Hint/examples/Hint.md]: @edonehoo I trimmed this down a little with my suggestions above, let me know what you think - open to further changes:
- https://patternfly-react-pr-10488.surge.sh/components/hint#with-actions

File: packages/react-core/src/components/Hint/examples/Hint.md [hunk 4]
```diff
@@ -20,3 +20,15 @@ import EllipsisVIcon from '@patternfly/react-icons/dist/esm/icons/ellipsis-v-ico
 ```ts file="HintBasicWithoutTitle.tsx"
 
 ```
+
+### With actions
+
+You can add actions to a hint to allow users to respond to suggestions and initiate any related processes.
+
+When a `<Hint>` contains `actions`, a negative margin is applied to the actions for proper alignment between the hint title and the actions. For larger-sized titles or non-plain actions such as a button, remove this negative margin by setting `hasNoActionsOffset` to `true`.
```
Comment (edonehoo) [inline packages/react-core/src/components/Hint/examples/Hint.md:28]:
```suggestion
When a `<Hint>` contains `actions`, a negative margin is applied to the actions so that they are better aligned with the hint title. For larger-sized titles or non-plain actions (such as a button), remove this negative margin by setting `hasNoActionsOffset` to `true`.
``` 
nitpicky but maybe this so there's one less "actions"?

File: packages/react-core/src/components/Hint/examples/Hint.md (hunk 2 above)
Comment (edonehoo) [reply packages/react-core/src/components/Hint/examples/Hint.md]: yes, great suggestion! I suggested a small tweak, but I also like it how it is

File: packages/react-core/src/components/Hint/examples/Hint.md (hunk 4 above)
Comment (evwilkin) [reply packages/react-core/src/components/Hint/examples/Hint.md:28]: Wow yes thanks, reading it back now all i see is the word "actions" so this change is worth it 👏

**Review by edonehoo — APPROVED** [review]

File: packages/react-core/src/components/Hint/__tests__/Hint.test.tsx [hunk 5]
```diff
@@ -65,3 +77,8 @@ test('matches hint snapshot', () => {
   const { asFragment } = render(<Hint />);
   expect(asFragment()).toMatchSnapshot();
 });
+
+test('actions are rendered', () => {
```
Comment (tlabaj) [inline packages/react-core/src/components/Hint/__tests__/Hint.test.tsx:81]: maybe verify here that the no offset style is not applied. Or add a new assertion.

Comment (evwilkin) [reply packages/react-core/src/components/Hint/__tests__/Hint.test.tsx:82]: @tlabaj I [updated this test](https://github.com/patternfly/patternfly-react/pull/10488/files#diff-7994ebe7a6d432372e22a19428e6e3c95f8b34afb095c707b7c55cb6d38fce5aR56), does that work?

Comment (tlabaj) [reply packages/react-core/src/components/Hint/__tests__/Hint.test.tsx:82]: Looks good!

**Review by tlabaj — APPROVED** [review]

**Comment by tlabaj** [comment]
Just needs a rebase.
